Tools/ssl/make_ssl_data.py

```diff
@@ -55,103 +109,178 @@ def _file_search(fname, pat):
+    """Generate table short libname to numeric code."""
+    yield "/* generated from args.lib2errnum */"
     yield "static struct py_ssl_library_code library_codes[] = {"
     for libname in sorted(args.lib2errnum):
         yield f"#ifdef ERR_LIB_{libname}"
         yield f'    {{"{libname}", ERR_LIB_{libname}}},'
         yield "#endif"
-    yield "    { NULL }"
+    yield "    {NULL, 0}  /* sentinel */"
     yield "};"
-    yield ""
 
 
 def gen_error_codes(args):
-    """Generate error code table for error reasons"""
+    """Generate error code table for error reasons."""
+    yield "/* generated from args.reasons */"
     yield "static struct py_ssl_error_code error_codes[] = {"
     for reason, libname, errname, num in args.reasons:
-        yield f"  #ifdef {reason}"
+        yield f"#ifdef {reason}"
         yield f'    {{"{errname}", ERR_LIB_{libname}, {reason}}},'
-        yield "  #else"
+        yield "#else"
         yield f'    {{"{errname}", {args.lib2errnum[libname]}, {num}}},'
-        yield "  #endif"
-
-    yield "    { NULL }"
+        yield "#endif"
+    yield "    {NULL, 0, 0}  /* sentinel */"
     yield "};"
-    yield ""
 
 
-def main():
-    args = parser.parse_args()
-
+def get_openssl_git_commit(args):
+    git_describe = subprocess.run(
+        ['git', 'describe', '--long', '--dirty'],
+        cwd=args.srcdir,
+        capture_output=True,
+        encoding='utf-8',
+        check=True,
+    )
+    # FEAT(picnixz): in the future, we may want to check that we only
+    # generate data from a Git commit that is a tag (to that end, we
+    # can either re-run 'git describe --long --dirty --contains' and
+    # compare the output (the HEAD case is special as it is never
+    # contained) or parse the output of 'git describe' to count
+    # the number of commits until the previous tag.
+    return git_describe.stdout.strip()
+
+
+def get_header_name(args, commit):
+    if args.openssl_version is not None:
+        return f"Py_SSL_DATA_{args.openssl_version}_H"
+    if args.output is not None:
+        # deduce the header from the filename
+        basename = os.path.basename(args.output)
+        filestem, _ = os.path.splitext(basename)
+        suffix = filestem.strip('_').upper()
+        return f"Py_SSL_DATA_{suffix.removeprefix('SSL_DATA_')}_H"
+    # deduce the header from the Git commit
+    match = re.match(r"openssl[^\d](\d+)[^\d](\d+)[^\d]([a-z0-9]+)",
+                     commit, re.ASCII | re.IGNORECASE)
+    if match is None:
+        error("cannot extract OpenSSL version from Git commit %r; "
+              "specify it via -H/--openssl-version instead", commit)
+    major, minor, patch = match.groups()
+    return f"Py_SSL_DATA_{major}{minor}{patch}_H"
+
+
+def gen_header_prolog_and_epilog(header_name):
+    prolog = f"""\
+#ifndef {header_name}
+#define {header_name}
+
+#include <stddef.h>
+
+#ifdef __cplusplus
+extern "C" {{
+#endif"""
+    epilog = f"""\
+#ifdef __cplusplus
```

Comment:
Why? We don't compile CPython with C++.
We also don't define the headers more than once (and can fix the issue if we somehow start to do that).

Reply:
In other words: These aren't reusable headers; they're more like chunks of C code, ripped out because they're autogenerated and conditional.

Reply:
I wanted the headers to look as close to other headers, but I can remove that "prett formatting"